storage/src/test/java/org/apache/kafka/storage/internals/log/LogSegmentTest.java

```diff
@@ -779,6 +784,75 @@ public void testDeleteIfExistsWithGetParentIsNull() throws IOException {
+        LogSegment segment = createSegment(0, 1, Time.SYSTEM);
+
+        ByteBuffer buffer1 = ByteBuffer.allocate(1024);
+        // append first batch to buffer1
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer1, Compression.NONE, TimestampType.CREATE_TIME, 0);
+        builder.append(0L, "key1".getBytes(), "value1".getBytes());
+        builder.close();
+
+        // append second batch to buffer1
+        builder = MemoryRecords.builder(buffer1, Compression.NONE, TimestampType.CREATE_TIME, 1);
+        builder.append(1L, "key1".getBytes(), "value1".getBytes());
+        builder.close();
+
+        buffer1.flip();
+        MemoryRecords record = MemoryRecords.readableRecords(buffer1);
+        segment.append(1L, record);
+
+        ByteBuffer buffer2 = ByteBuffer.allocate(1024);
+        // append first batch to buffer2
+        builder = MemoryRecords.builder(buffer2, Compression.NONE, TimestampType.CREATE_TIME, 2);
+        builder.append(2L, "key1".getBytes(), "value1".getBytes());
+        builder.close();
+
+        buffer2.flip();
+        record = MemoryRecords.readableRecords(buffer2);
+        segment.append(2L, record);
+
+        assertEquals(2, segment.offsetIndex().entries());
+        assertTrue(segment.offsetIndex().lookup(2L).position > segment.offsetIndex().lookup(1L).position);
+
+        assertEquals(2, segment.timeIndex().entries());
+        assertTrue(segment.timeIndex().lookup(2L).offset > segment.timeIndex().lookup(1L).offset);
+    }
+
+    @Test
+    public void testNonMonotonicTimestampForMultipleBatchesInMemoryRecords() throws IOException {
+        LogSegment segment = createSegment(0, 1, Time.SYSTEM);
+
+        ByteBuffer buffer1 = ByteBuffer.allocate(1024);
+        // append first batch to buffer1
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer1, Compression.NONE, TimestampType.CREATE_TIME, 0);
+        builder.append(1L, "key1".getBytes(), "value1".getBytes());
+        builder.close();
+
+        // append second batch to buffer1
+        builder = MemoryRecords.builder(buffer1, Compression.NONE, TimestampType.CREATE_TIME, 1);
+        builder.append(0L, "key1".getBytes(), "value1".getBytes());
+        builder.close();
+
+        // append third batch to buffer1
+        builder = MemoryRecords.builder(buffer1, Compression.NONE, TimestampType.CREATE_TIME, 2);
+        builder.append(2L, "key1".getBytes(), "value1".getBytes());
+        builder.close();
+
+        buffer1.flip();
+        MemoryRecords record = MemoryRecords.readableRecords(buffer1);
+        segment.append(2L, record);
+
+        assertEquals(2, segment.offsetIndex().entries());
+        assertEquals(1, segment.offsetIndex().lookup(1L).offset);
+        assertEquals(2, segment.offsetIndex().lookup(2L).offset);
+
+        assertEquals(1, segment.timeIndex().entries());
+        assertEquals(0L, segment.timeIndex().lookup(1L).offset);
+        assertEquals(2L, segment.timeIndex().lookup(2L).offset);
```

Comment:
Could we use `timeIndex().entry` to explicitly fetch each index entry instead of using `timeIndex().lookup`?

Reply:
Updated assertion. Thanks.